content/specs/auth.md

```diff
@@ -0,0 +1,390 @@
+
+The client assertion type to use is `urn:ietf:params:oauth:client-assertion-type:jwt-bearer`, as described in "JSON Web Token (JWT) Profile for OAuth 2.0 Client Authentication and Authorization Grants" ([RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523)). Clients and Authorization Servers must support the `ES256` cryptographic system. The set of recommended systems/algorithms is expected to evolve over time.
+
+Additional requirements:
+
+- confidential clients publish one or more client authentication keys (public key) in the client metadata. This can be either direct JWK format as JSON in the `jwks` field, or as a separate JWKS JSON object on the web linked by a `jwks_uri` URL. A `jwks_uri` URL must be a valid fully qualified URL with `https://` scheme.
+- confidential clients should periodically rotate client keys, adding new keys to the JWKS set and using then for new sessions, then removing old keys once they are no longer associated with any active auth sessions
+- confidential clients must include `token_endpoint_auth_method` as `private_key_jwt` in their client metadata document
+- confidential clients are expected to immediately remove client authentication keys from their client metadata if they key has been leaked or compromised
+- Authorization Servers must bind active auth sessions for confidential clients to the client authentication key used at the start of the session. The server should revoke the session and reject further token refreshes if the client authentication key becomes absent from the client metadata. This means the Authorization Server is expected to periodically re-fetch client metadata.
+
+TODO: what if the auth server fails to fetch the client metadata at all? eg a transient TCP connection failure, or server is just down for a couple hours. seems a bit brittle to immediately revoke, maybe a grace period?
+
+### Tokens and Session Lifetime
+
+Access tokens are used to authorize client requests to the account's PDS (”Resource Server”). From the standpoint of the client, they are opaque, but they are often signed JWTs including an expiration time. Depending on the PDS implementation, it may or may not be possible to revoke individual access tokens in the event of a compromise, so they must be restricted relatively short lifetime.
+
+Refresh tokens are used to request new tokens (of both types) from the Authorization Server (PDS or entryway). They are also opaque from the standpoint of clients. Auth sessions can be revoked, invalidating the refresh tokens, so they may have longer lifetime. In the atproto OAuth profile, refresh tokens are generally single-use, with the "new" refresh token replacing that used in the token request. This means client implementations may need locking primitives to prevent concurrent token refresh requests.
```

Comment:
```suggestion
Refresh tokens are used to request new tokens (of both types) from the Authorization Server (PDS or entryway). They are also opaque from the standpoint of clients. Auth sessions can be revoked - invalidating the refresh tokens - so they may have longer lifetime. In the atproto OAuth profile, refresh tokens are generally single-use, with the "new" refresh token replacing that used in the token request. This means client implementations may need locking primitives to prevent concurrent token refresh requests.
```